{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ecc6a985-29a7-44e4-8f27-c6a8883f3a41",
   "metadata": {},
   "source": [
    "# Nichesphere differential co-localization tutorial\n",
    "\n",
    "Nichesphere is an sc-verse compatible Python library which allows the user to find differentially co-localized cellular niches and biological processes involved in their interactions based on cell type pairs co-localization probabilities in different conditions. Cell type pair co-localization probabilities are obtained in different ways: from deconvoluted Visium 10x / PIC-seq data (probabilities of finding each cell type in each spot / multiplet), or counting cell boundaries overlaps for each cell type pair in single cell spatial data (MERFISH , CODEX ...). This tutorial focuses on defining groups of cells that converge or split in disease (Ischemia) based on differential co-localization. \n",
    "\n",
    "Nichesphere also offers the possibility to look at localized differential cell - cell communication based on Ligand-Receptor pairs expression data, such as results from CrossTalkeR [ref]. This is addressed in the localized differential communication tutorial.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "099b15cb-54b9-45bf-a1ab-a0cd308d629d",
   "metadata": {},
   "source": [
    "## 1. Libraries and functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "994efbef-4ac0-4c33-9557-3d4fa434793e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scipy\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as colors\n",
    "import networkx as nx\n",
    "import warnings\n",
    "import scanpy as sc\n",
    "import mudata as md\n",
    "import numpy as np\n",
    "from community_layout.layout_class import CommunityLayout\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import os\n",
    "import math\n",
    "\n",
    "import nichesphere"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35ad2170-35c7-439f-9ba1-5b8f909002ab",
   "metadata": {},
   "source": [
    "## 2. Data at first glance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bfdabd1-fec7-4f67-9b46-e77153a2692f",
   "metadata": {},
   "source": [
    "In this example we will use data from the Myocardial Infarction atlas from Kuppe, C. et. Al., 2022"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56909253-2afc-4548-b15b-5b376518ef67",
   "metadata": {},
   "outputs": [],
   "source": [
    "mudata=md.read('heart_MI_ST_SC_23samples.h5mu')\n",
    "mudata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93555b4e-3f23-492d-aa02-a33323e65cb8",
   "metadata": {},
   "source": [
    "This is a subset with 23 samples (samples with less than 1500 cells in the snRNA-seq data were filtered out), and 33 different cell subtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2de30ee5-3b20-43f3-bea1-c0c33b44ba93",
   "metadata": {},
   "outputs": [],
   "source": [
    "mudata['sc'].obsm['umap']=mudata['sc'].obsm['X_umap_harmony']\n",
    "sc.pl.umap(mudata['sc'], \n",
    "           color=['cell_type', 'cell_subtype2'], wspace=0.3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b744a399-b5aa-435e-9ebd-3a8389f8d9b8",
   "metadata": {},
   "source": [
    "In this case, we will get cell type co-localization probabilities from **deconvoluted Visium** data (Cell type probabilities per spot): \n",
    "\n",
    "In a previous step, we used MOSCOT(Klein et. al., 2025) to deconvolute cell subtypes in visium slices from the same 23 samples , getting matrices of probabilities of each cell being in each spot. Then we got cell type probabilities per spot summing the probabilities of cells of the same kind in each spot; thus getting cell type probability matrices for all samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae3493a1-673e-4ec2-9588-1709da1d3373",
   "metadata": {},
   "source": [
    "(you can have a closer look at these steps in the preprocessing tutorial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca576d97-59fa-4dc7-bc60-b5618d601717",
   "metadata": {},
   "outputs": [],
   "source": [
    "CTprops=pd.read_csv('CTprops.csv', index_col=0)\n",
    "CTprops.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "486f7916-a1ce-490c-b172-c4e53ceee5ac",
   "metadata": {},
   "source": [
    "From these deconvolution results, we can look at **cell type proportions per sample**. For this we will need the spot ID and sample correspondence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98c3f35b-e8fe-4ef7-8a7e-bb79a3ea845b",
   "metadata": {},
   "outputs": [],
   "source": [
    "spotSamples=mudata['visium'].obs.patient_region_id\n",
    "spotSamples.reset_index().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "558eaecd-b7ce-41c9-9d02-11e1d02406e9",
   "metadata": {},
   "source": [
    "A way to check the deconvolution proportions is using a clustermap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbd25da0-42b7-40c4-b4a7-1b580334ce83",
   "metadata": {},
   "outputs": [],
   "source": [
    "CTprops_sample=CTprops.copy()\n",
    "CTprops_sample['sample']=mudata['visium'].obs.patient_region_id\n",
    "sns.clustermap(CTprops_sample.groupby('sample').sum().T/CTprops_sample.groupby('sample').sum().sum(axis=1) , \n",
    "               cmap='Blues', method='ward')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49691fa0-3feb-4911-ba47-5df83770b64c",
   "metadata": {},
   "source": [
    "Alternativelly, we can check the deconvolution proportions using barplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41f1ba93-8da0-4e39-abf9-81b4dece88e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "t1=pd.DataFrame(CTprops[spotSamples=='control_P7'].sum(), columns=['control_P7'])\n",
    "t2=pd.DataFrame(CTprops[spotSamples=='IZ_P15'].sum(), columns=['IZ_P15'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f1435e9-9ed4-44f4-929f-23f3ccb7be7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale=1)\n",
    "sns.set_style(rc = {'axes.facecolor': 'white'})\n",
    "\n",
    "fig, axes = plt.subplots(1, 2, figsize=(20, 7))\n",
    "\n",
    "sns.barplot(ax=axes[0], y=t1.sort_values('control_P7', ascending=False).index, x='control_P7', \n",
    "            data=t1.sort_values('control_P7', ascending=False), color='darkblue')\n",
    "axes[0].set_title('control_P7')\n",
    "\n",
    "sns.barplot(ax=axes[1], y=t2.sort_values('IZ_P15', ascending=False).index, x='IZ_P15',\n",
    "            data=t2.sort_values('IZ_P15', ascending=False), color='darkred')\n",
    "axes[1].set_title('IZ_P15')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "accedf55-2269-4e42-aeaa-c94d213a73f4",
   "metadata": {},
   "source": [
    "We can visualize **cell type deconvolution results in slices** (spots are colored by the the cell type with highest proportion). For this we will need the spatial coordinates of the spots in the Visium slices need to be in the slot **uns['spatial']** of the Visium anndata object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c016009-dcbc-48d5-a445-895e791e0113",
   "metadata": {},
   "outputs": [],
   "source": [
    "mudata['visium'].uns['spatial']=mudata['visium'].obsm['X_spatial']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a90af9ff-9eff-4d27-8796-854a9bcd0ee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "idPat = 'GT_IZ_P9'\n",
    "nichesphere.coloc.spatialCTPlot(adata=mudata['visium'][mudata['visium'].obs.patient_region_id==idPat].copy(), \n",
    "                                CTprobs=CTprops.loc[spotSamples.index[spotSamples==idPat]], \n",
    "                                cell_types=mudata['sc'].obs.cell_subtype2, spot_size=0.015, \n",
    "                                legend_fontsize=7)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db12f15c-6888-4988-9abd-806d18f75b68",
   "metadata": {},
   "source": [
    "## 3. Co-localization\n",
    "\n",
    "We computed then co-localization probabilities from the cell type probability matrices. Here we got concatenated co-localization sample matrices of cell type x cell type.\n",
    "\n",
    "Then we reshaped the co-localization data into a matrix of cell type pairs x samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30b3f700-d12a-46c9-9de4-d6bc564420ba",
   "metadata": {},
   "source": [
    "(you can have a closer look at these steps in the preprocessing tutorial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e829b9d8-7abd-4bc0-b080-5d3aa3c424e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "colocPerSample=pd.read_csv('colocPerSample.csv', index_col=0)\n",
    "colocPerSample.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9b08e3a-a3b1-4417-a6ef-2897b21e73b1",
   "metadata": {},
   "source": [
    "The sum of the probabilities of every cell type pair in a sample must be = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e786214-cbd5-41ef-b23c-a14cac58a6c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "colocPerSample.sum(axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acd47545-5430-48e7-957c-c5169d4decc2",
   "metadata": {},
   "source": [
    "Same cell type interactions will be excluded later on, so we'll have a list of same cell type interaction pairs in order to subset the co-localization table we'll generate in the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8ecab64-cda9-4ca8-a159-df6e02f86b8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types=CTprops.columns\n",
    "oneCTints=cell_types+'-'+cell_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64ea88ce-fbee-4c88-acda-b794331cf228",
   "metadata": {},
   "source": [
    "**Conditions**\n",
    "\n",
    "We will need the following metadata to subset the samples in **control (myogenic)** and **disease (ischemic)**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53e0f246-fe34-4eff-843a-5b9a200d19eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleTypesDF=pd.read_csv('MI_sampleTypesDF.csv')\n",
    "sampleTypesDF.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c99de8b-5943-4707-bce2-748be20a4f39",
   "metadata": {},
   "source": [
    "## 4. Differential co-localization analysis\n",
    "\n",
    "We will test differential co-localization between **myogenic** and **ischemic** samples using Wilcoxon rank sums tests:\n",
    "\n",
    "**Null Hypothesis (H0):**\n",
    "The median of the population of differences between co-localization probabilities of cell types a and b in myogenic and ischemic samples is zero.\n",
    "\n",
    "**Alternative Hypothesis (H1):**\n",
    "The median of the population of differences between co-localization probabilities of cell types a and b in myogenic and ischemic samples is not equal to zero."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f98ac299-0768-47ff-a0af-72bc0b2b3fbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "myo_iscDF=nichesphere.coloc.diffColoc_test(coloc_pair_sample=colocPerSample, \n",
    "                                           sampleTypes=sampleTypesDF, \n",
    "                                           exp_condition='ischemic', \n",
    "                                           ctrl_condition='myogenic')\n",
    "myo_iscDF.head() # myp_iscDF type: pd.DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b765783-9787-4f99-b8cf-9cf687afd614",
   "metadata": {},
   "source": [
    "Then we will reshape the data to visualize the Wilcoxon test scores in a heatmap and filter non significant co-localization differences using the parameter **p** (in this case, scores with p-values > 0.1 are filtered out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbcbe078-d5ad-45ac-a42e-2e55268af5a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "myo_isc_HM=nichesphere.tl.pval_filtered_HMdf(testDF=myo_iscDF, \n",
    "                                             oneCTinteractions=oneCTints, \n",
    "                                             p=0.1,                             #threshold p-value to filter\n",
    "                                             cell_types=cell_types)\n",
    "myo_isc_HM.head() # myo_isc_HM type: pd.DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f312b23-fe05-4f6a-8b12-67a575e3a9ad",
   "metadata": {},
   "source": [
    "As the cells classified as proliferating cells (prolif) are many different cell types and thus hard to interpret, we'll remove them for further analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a123cb13-8d6c-4299-a636-153852a24844",
   "metadata": {},
   "outputs": [],
   "source": [
    "myo_isc_HM=myo_isc_HM.loc[myo_isc_HM.columns.str.contains('prolif')==False,myo_isc_HM.index.str.contains('prolif')==False]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ddb7118-d6c0-4896-9255-1e37105e41ff",
   "metadata": {},
   "source": [
    "We will also remove cells with no significant co-localization differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2b4bfbf-9433-49b7-83fa-0bbd3edd4faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "myo_isc_HM=myo_isc_HM.loc[myo_isc_HM.sum()!=0,myo_isc_HM.sum()!=0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cca781d9-a14d-4b0d-8117-c7b42498a84d",
   "metadata": {},
   "source": [
    "Now we can plot the differential co-localization scores heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72cefa68-49f7-4a42-a92a-05052db3b223",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale=1)\n",
    "plot=sns.clustermap(myo_isc_HM, cmap='vlag', center=0, method='ward', cbar_kws={'label': 'diffColoc. Score'})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e59b458a-8586-4fbb-893c-e8570fe7a817",
   "metadata": {},
   "source": [
    "**Differential co-localization network**\n",
    "\n",
    "To build the differential co-localization network, we will get an **adjacency matrix** (adj) based on the **euclidean distances** among the distributions of significant differential co-localization scores for the different cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "775ecbff-c2c8-452d-a0aa-e4f738f11c6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "HMdist=pd.DataFrame(scipy.spatial.distance.squareform(scipy.spatial.distance.pdist(myo_isc_HM)), \n",
    "                    columns=myo_isc_HM.columns, index=myo_isc_HM.index)\n",
    "\n",
    "HMsimm=1-HMdist/HMdist.max().max()\n",
    "##Cell pairs with not significant differential co-localization get 0\n",
    "HMsimm[myo_isc_HM==0]=0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62f77aa3-894a-4cf9-a243-4b628e1e876b",
   "metadata": {},
   "source": [
    "A **cell group dictionary** should be used here to visualize different cell groups in different colors. As we don't have cell groups yet, we'll have a dictionary of all cells in one group and a list of one color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10f611fb-2b51-4ac3-b09c-5fe444afafcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "niches_dict={'1_': list(myo_isc_HM.index)}\n",
    "clist=['#4daf4a']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78eb6625-c82e-482e-80a1-6eddfcfe92b9",
   "metadata": {},
   "source": [
    "Now we can plot the differential co-localization network using the **colocNW** function from Nichesphere. This function has many parameters that can be tuned: \n",
    "\n",
    "**nodeSize** for example, defines how the size of the nodes will be calculated. Options are 'betweeness', 'pagerank' (both network statistics) and None (all nodes have the same size).\n",
    "**alpha** indicates the transparency of the edges and in goes from 0 (completely transparent) to 1 (opaque)\n",
    "**fsize** is the size of the figure (x,y)\n",
    "\n",
    "This function returns the network with the edge weights corresponding to the diff. coloc. scores (positive and negative)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb2c090b-bc54-4dcf-8f71-5170468393ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['axes.facecolor'] = \"None\"\n",
    "nichesphere.coloc.colocNW(x_diff=myo_isc_HM,            #differential co-localization matrix\n",
    "                          adj=HMsimm,                   #adjacency matrix\n",
    "                          cell_group=niches_dict, \n",
    "                          clist=clist, \n",
    "                          nodeSize='betweeness',        \n",
    "                          lab_spacing=9,                #space between node and label\n",
    "                          alpha=0.4,                    #edges transparency\n",
    "                          fsize=(12,12))                #figure size"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81870d7b-ae8a-42f9-afaa-b48d2d7d937a",
   "metadata": {},
   "source": [
    "Now we'll do community detection using Louvain. First we will get the network from the adjacency matrix as we won't use the signed weights for this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "658a9589-e78b-491b-916c-fb0f75faf1d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gCol_unsigned=nx.from_pandas_adjacency(HMsimm, create_using=nx.Graph)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30e322cf-597f-4952-8c06-e30e12de14e1",
   "metadata": {},
   "source": [
    "We will use the community-layout library function **CommunityLayout** to show the communities in a layout suited for this. This function is compatible with networkx (Hagberg et. al., 2008) community detection functions, which will be used internally as indicated by the parameters **community_algorithm** and **community_kwargs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24dad325-8d69-4ec5-b7a9-358f6477b913",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate community layout\n",
    "cl=CommunityLayout(gCol_unsigned,\n",
    "        community_compression = 0.4,\n",
    "        layout_algorithm = nx.spring_layout,\n",
    "        layout_kwargs = {\"k\":75, \"iterations\":1000},\n",
    "        community_algorithm = nx.algorithms.community.louvain_communities,\n",
    "        community_kwargs = {\"resolution\":1.1,  'seed':12, 'weight':'weight'})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48e28a7b-217a-42fd-871f-89554208c5b3",
   "metadata": {},
   "source": [
    "We can extract the communities (niches) as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa52bcf9-d1ff-42d0-88d3-b5386923e997",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = {index: list(value) for index, value in enumerate(cl.communities())}\n",
    "print(pd.DataFrame.from_dict(d, orient='index').T.to_string(index=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed76253-2809-42e1-a90b-2e6976d11f4d",
   "metadata": {},
   "source": [
    "And then name them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "397c4faf-4d72-45af-9d8d-145890f39dfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "niche_names=['1_Stromal', '2_Stressed_CM', '3_Healthy_CM', '4_Fibrotic']\n",
    "niches_dict=dict(zip(niche_names,list(d.values()))) \n",
    "print(pd.DataFrame.from_dict(niches_dict, orient='index').T.to_string(index=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e1dd4b0-d7a3-4e89-a207-937bfc898e56",
   "metadata": {},
   "source": [
    "And assign them colors to color the network nodes according to their niche"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7ee6d56-b7e1-477b-92d2-34aae744a32d",
   "metadata": {},
   "outputs": [],
   "source": [
    "clist=['#4daf4a', '#0072B5', '#BC3C29', '#ffff33']\n",
    "niche_cols=pd.Series(clist, index=list(niches_dict.keys()))\n",
    "niches_df=nichesphere.tl.cells_niche_colors(CTs=CTprops.columns, \n",
    "                                            niche_colors=niche_cols, \n",
    "                                            niche_dict=niches_dict)\n",
    "niches_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c832201e-b235-4314-a694-c5d4b7f53ede",
   "metadata": {},
   "source": [
    "Then we can get the node positions to input them to the nichesphere **colocNW** function through the **pos** parameter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e13b2cbf-a466-40bd-8f29-84c7274970bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos=cl.full_positions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c907114c-3ab2-47c8-b679-0a29e6315b02",
   "metadata": {},
   "source": [
    "And plot the niches on the community layout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf0b843d-def6-48ef-aa8c-5c1a8e21d446",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['axes.facecolor'] = \"None\"\n",
    "\n",
    "gCol=nichesphere.coloc.colocNW(x_diff=myo_isc_HM, \n",
    "                               adj=HMsimm,\n",
    "                               cell_group=niches_dict, \n",
    "                               clist=clist, \n",
    "                               nodeSize='betweeness', \n",
    "                               layout=None,                         #layout needs to be set to None if we provide node positions\n",
    "                               lab_spacing=0.05, \n",
    "                               thr=1, \n",
    "                               alpha=0.4, \n",
    "                               fsize=(10,10), \n",
    "                               pos=pos,                             #node positions (from the CommunityLayout function)\n",
    "                               edge_scale=1,                        #edge width\n",
    "                               legend_ax=[0.7, 0.05, 0.15, 0.2])    #legend position\n",
    "#Legend\n",
    "legend_elements1=[plt.Line2D([0], [0], marker=\"o\" ,color='w', markerfacecolor=clist[i], lw=4, \n",
    "                             label=list(niches_dict.keys())[i], ms=10) for i in range(len(list(niches_dict.keys())))]\n",
    "plt.gca().add_artist(plt.legend(handles=legend_elements1,loc='lower left', fontsize=13, title='Niches', \n",
    "                                alignment='left'))\n",
    "#plt.savefig('diffColocNW_CD.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60a17737",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6b046433",
   "metadata": {},
   "source": [
    "### Network Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb76c83e-d7f4-457c-8c50-69935bf6fcdf",
   "metadata": {},
   "source": [
    "We can also calculate some network statistics with the networkx package functions (this will be done on the signed network):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d71b4e44-b162-4a22-8702-72f3c5d91c65",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0=pd.DataFrame({'betweenness':[nx.betweenness_centrality(gCol)[x] for x in list(gCol.nodes)], \n",
    "                 'degree':[nx.degree_centrality(gCol)[x] for x in list(gCol.nodes)], \n",
    "                 'pagerank':[nx.pagerank(gCol, weight=None)[x] for x in list(gCol.nodes)]})\n",
    "t0.index=list(gCol.nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b12e7921-892a-4aae-8d17-73d32e68453f",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 3, figsize=(21, 7))\n",
    "for i in range(len(t0.columns)):\n",
    "    _ = sns.barplot(ax=axes[i], y=t0.sort_values(t0.columns[i], ascending=False).index[0:5], x=t0.columns[i], \n",
    "        data=t0.sort_values(t0.columns[i], ascending=False)[0:5], color='purple')\n",
    "    axes[i].set_title(t0.columns[i])\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4df33567-98c2-4e91-8713-f645e640a9ed",
   "metadata": {},
   "source": [
    "So we can look separately at positive and negative degree:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea8f41a7-d224-4388-aa27-4900cd67b959",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Positive edges stats\n",
    "G_pos=gCol.copy()\n",
    "to_remove=[(a,b) for a, b, attrs in G_pos.edges(data=True) if attrs[\"weight\"] <= 0]\n",
    "G_pos.remove_edges_from(to_remove)\n",
    "\n",
    "t1=pd.DataFrame({'degree':[nx.degree_centrality(G_pos)[x] for x in list(G_pos.nodes)]})\n",
    "t1.index=list(G_pos.nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af8f7781-3a97-48c3-a7b8-e276e0f042be",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Negative edges stats\n",
    "G_neg=gCol.copy()\n",
    "to_remove=[(a,b) for a, b, attrs in G_neg.edges(data=True) if attrs[\"weight\"] >= 0]\n",
    "G_neg.remove_edges_from(to_remove)\n",
    "\n",
    "t2=pd.DataFrame({'degree':[nx.degree_centrality(G_neg)[x] for x in list(G_neg.nodes)]})\n",
    "t2.index=list(G_neg.nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1c330af-46e6-41af-9a4b-388ebce113ce",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 2, figsize=(14, 7))\n",
    "\n",
    "_=sns.barplot(ax=axes[0], y=t1.sort_values('degree', ascending=False).index[0:5], x='degree', \n",
    "              data=t1.sort_values('degree', ascending=False)[0:5], color='red')\n",
    "axes[0].set_title('degree_positive')\n",
    "\n",
    "_=sns.barplot(ax=axes[1], y=t2.sort_values('degree', ascending=False).index[0:5], x='degree', \n",
    "              data=t2.sort_values('degree', ascending=False)[0:5], color='blue')\n",
    "axes[1].set_title('degree_negative')\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b263ffc1-10a6-484c-ba6f-cb7e8927aa7d",
   "metadata": {},
   "source": [
    "caculate betweeness for positive and negative subgraph seperately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9756a04-7c40-4702-bfe0-9df67409e88c",
   "metadata": {},
   "outputs": [],
   "source": [
    "betw_pos = pd.DataFrame({'betweenness_pos':[nx.betweenness_centrality(G_pos)[x] for x in list(G_pos.nodes)]})\n",
    "betw_pos.index=list(G_pos.nodes)\n",
    "\n",
    "betw_neg = pd.DataFrame({'betweenness_neg':[nx.betweenness_centrality(G_neg)[x] for x in list(G_neg.nodes)]})\n",
    "betw_neg.index=list(G_neg.nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7297c14",
   "metadata": {},
   "source": [
    "#### Prestige"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcbde3ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def node_prestige(node, pos_degree, neg_degree):\n",
    "    \"\"\"\n",
    "    calculate prestige of a node\n",
    "\n",
    "    :pos_degree: degree of node in network for positive links\n",
    "    :neg_degree: degree of node in network for negative links\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    pres = ( abs(pos_degree) - abs(neg_degree) ) / ( abs(pos_degree) + abs(neg_degree) )\n",
    "    return pres\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4833d230",
   "metadata": {},
   "outputs": [],
   "source": [
    "prestige = pd.DataFrame({'prestige':[node_prestige(node,t1['degree'][node],t2['degree'][node]) for node in list(gCol.nodes)]})\n",
    "prestige.index = list(gCol.nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e27bc7d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### remove all entries with a prestige of -1. for better overview for plotting\n",
    "prestige_removed = {key:val for key, val in prestige['prestige'].items() if val != -1.}\n",
    "df_prestige = pd.DataFrame.from_dict({'prestige':prestige_removed})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5043d20",
   "metadata": {},
   "outputs": [],
   "source": [
    "### plot positive and negative prestige scores\n",
    "fig, axes = plt.subplots(1, 2, figsize=(14, 7))\n",
    "\n",
    "sns.barplot(ax=axes[0], y=df_prestige.sort_values('prestige', ascending=False).index[0:4], x='prestige', \n",
    "              data=df_prestige.sort_values('prestige', ascending=False)[0:4], color='red')\n",
    "axes[0].set_title('prestige_positive')\n",
    "\n",
    "sns.barplot(ax=axes[1], y=df_prestige.sort_values('prestige', ascending=True).index[0:6], x='prestige', \n",
    "              data=df_prestige.sort_values('prestige', ascending=True)[0:6], color='blue')\n",
    "axes[1].set_title('prestige_negative')\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee6d903e",
   "metadata": {},
   "source": [
    "Plot sign split degree and prestige"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f08d72c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 3, figsize=(21, 7))\n",
    "#sorted_y_order = prestige.sort_values('prestige', ascending=True).index\n",
    "sorted_y_order = t2.sort_values('degree', ascending=False).index\n",
    "\n",
    "sns.barplot(ax=axes[0], y=sorted_y_order, x='degree', \n",
    "              data=t2.loc[sorted_y_order], color='blue')\n",
    "axes[0].set_title('degree_negative')\n",
    "\n",
    "sns.barplot(ax=axes[1], y=sorted_y_order, x='degree', \n",
    "              data=t1.loc[sorted_y_order], color='red')\n",
    "axes[1].set_title('degree_positive')\n",
    "\n",
    "sns.barplot(ax=axes[2], y=sorted_y_order, x='prestige', \n",
    "              data=prestige.loc[sorted_y_order], color='purple')\n",
    "axes[2].set_title('prestige')\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c3b21b3",
   "metadata": {},
   "source": [
    "#### Signed PageRank"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10334032",
   "metadata": {},
   "source": [
    "Approach: calculate PageRank for positive and negative network seperatly"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c94e394-5311-4832-92e2-381bcb7df7da",
   "metadata": {},
   "source": [
    "Reference: X. Yin, X. Hu, Y. Chen, X. Yuan, and B. Li, “Signed-PageRank: An Efficient Influence Maximization Framework for Signed Social Networks,” IEEE Trans. Knowl. Data Eng., vol. 33, no. 5, pp. 2208–2222, May 2021, doi: 10.1109/TKDE.2019.2947421."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "979dbf5e-efbf-4ddd-a778-b1d1a2c72c35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge results of previous calculations\n",
    "results = pd.DataFrame()\n",
    "results[\"degree\"] = t0[\"degree\"]\n",
    "results[\"degree_pos\"] = t1[\"degree\"]\n",
    "results[\"degree_neg\"] = t2[\"degree\"]\n",
    "results[\"prestige\"] = prestige[\"prestige\"]\n",
    "results[\"betweenness_pos\"] = betw_pos[\"betweenness_pos\"]\n",
    "results[\"betweenness_neg\"] = betw_neg[\"betweenness_neg\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88a0be6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# split network into subgraphs with only positive and only negative edges\n",
    "G_pos = nx.DiGraph()\n",
    "G_neg = nx.DiGraph()\n",
    "\n",
    "for u, v, data in gCol.edges(data=True):\n",
    "    weight = data.get('weight', 1)\n",
    "    if weight > 0:\n",
    "        G_pos.add_edge(u, v, weight=weight)\n",
    "    elif weight < 0:\n",
    "        G_neg.add_edge(u, v, weight=abs(weight)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10ba6495",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate PageRank for positive and negative edges\n",
    "PR_pos=pd.DataFrame({'pagerank_pos':[nx.pagerank(G_pos, weight='weight')[x] for x in list(G_pos.nodes)]})\n",
    "PR_pos.index=list(G_pos.nodes)\n",
    "\n",
    "PR_neg=pd.DataFrame({'pagerank_neg':[nx.pagerank(G_neg, weight='weight')[x] for x in list(G_neg.nodes)]})\n",
    "PR_neg.index=list(G_neg.nodes)\n",
    "\n",
    "# align index of pos and neg DataFrame\n",
    "PR_pos = PR_pos.reindex(PR_neg.index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73e801cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# store results in DataFrame\n",
    "results[\"pagerank_pos\"] = PR_pos[\"pagerank_pos\"]\n",
    "results[\"pagerank_neg\"] = PR_neg[\"pagerank_neg\"]\n",
    "\n",
    "# calculate signed PageRanks as difference of positive and negative PageRank\n",
    "results[\"pagerank_signed\"] = PR_pos[\"pagerank_pos\"].fillna(0) - PR_neg[\"pagerank_neg\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5207537b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 3, figsize=(21, 7))\n",
    "sorted_y_order = PR_signed[\"pagerank_signed\"].sort_values(ascending=False).index\n",
    "\n",
    "sns.barplot(ax=axes[0], y=sorted_y_order, x='pagerank_neg', \n",
    "              data=PR_signed.loc[sorted_y_order], color='blue')\n",
    "axes[0].set_title('pagerank_neg')\n",
    "\n",
    "sns.barplot(ax=axes[1], y=sorted_y_order, x='pagerank_pos', \n",
    "              data=PR_signed.loc[sorted_y_order], color='red')\n",
    "axes[1].set_title('pagerank_pos')\n",
    "\n",
    "sns.barplot(ax=axes[2], y=sorted_y_order, x='pagerank_signed', \n",
    "              data=PR_signed.loc[sorted_y_order], color='purple')\n",
    "axes[2].set_title('pagerank_signed')\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f4feb48",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f43bc380-3c27-4706-8203-a2e09659b3e4",
   "metadata": {},
   "source": [
    "Unsigned and signed PageRank vs. unsigned degree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c14b85bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 3, figsize=(21, 7))\n",
    "sorted_y_order = results[\"pagerank_signed\"].sort_values(ascending=False).index\n",
    "\n",
    "sns.barplot(ax=axes[0], y=sorted_y_order, x='pagerank_signed', \n",
    "              data=results.loc[sorted_y_order], color='blue')\n",
    "axes[0].set_title('pagerank_signed')\n",
    "\n",
    "sns.barplot(ax=axes[1], y=sorted_y_order, x='pagerank_unsigned', \n",
    "              data=results.loc[sorted_y_order], color='red')\n",
    "axes[1].set_title('pagerank_unsigned')\n",
    "\n",
    "sns.barplot(ax=axes[2], y=sorted_y_order, x='degree', \n",
    "              data=results.loc[sorted_y_order], color='purple')\n",
    "axes[2].set_title('degree')\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca6425b9-f06d-46cf-8324-b6b20815edaf",
   "metadata": {},
   "source": [
    "Unsigned and signed PageRank vs. positive and negative Degree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3e179e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 4, figsize=(28, 7))\n",
    "sorted_y_order = results[\"pagerank_signed\"].sort_values(ascending=False).index\n",
    "\n",
    "sns.barplot(ax=axes[0], y=sorted_y_order, x='pagerank_signed', \n",
    "              data=results.loc[sorted_y_order], color='purple')\n",
    "axes[0].set_title('pagerank_signed')\n",
    "\n",
    "sns.barplot(ax=axes[1], y=sorted_y_order, x='pagerank_unsigned', \n",
    "              data=results.loc[sorted_y_order], color='green')\n",
    "axes[1].set_title('pagerank_unsigned')\n",
    "\n",
    "sns.barplot(ax=axes[2], y=sorted_y_order, x='degree_neg', \n",
    "              data=results.loc[sorted_y_order], color='blue')\n",
    "axes[2].set_title('degree_negative')\n",
    "\n",
    "sns.barplot(ax=axes[3], y=sorted_y_order, x='degree_pos', \n",
    "              data=results.loc[sorted_y_order], color='red')\n",
    "axes[3].set_title('degree_positive')\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "874e7109",
   "metadata": {},
   "outputs": [],
   "source": [
    "results[\"delta\"] = results[\"pagerank_signed\"] - results[\"pagerank_unsigned\"]\n",
    "\n",
    "plt.figure(figsize=(10, 8))\n",
    "plt.scatter(\n",
    "    results[\"pagerank_unsigned\"], \n",
    "    results[\"pagerank_signed\"],\n",
    "    c=results[\"delta\"],\n",
    "    cmap=\"coolwarm\",  # blue = negative delta, red = positive\n",
    "    edgecolor='k'\n",
    ")\n",
    "\n",
    "for idx, row in results.iterrows():\n",
    "    plt.text(row[\"pagerank_unsigned\"] + 0.001, row[\"pagerank_signed\"] + 0.001, idx, fontsize=8)\n",
    "\n",
    "plt.axhline(0, color='gray', linestyle='--')\n",
    "plt.axvline(0, color='gray', linestyle='--')\n",
    "plt.xlabel(\"Unsigned PageRank\")\n",
    "plt.ylabel(\"Signed PageRank\")\n",
    "plt.title(\"Signed vs Unsigned PageRank\")\n",
    "plt.colorbar(label=\"Delta (signed - unsigned)\")\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5d950da",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(2, 2, figsize=(14, 10))\n",
    "\n",
    "# Normalize the colormap across both plots\n",
    "norm = colors.Normalize(vmin=results['pagerank_signed'].min(), vmax=results['pagerank_signed'].max())\n",
    "cmap = cm.coolwarm\n",
    "\n",
    "# Plot 1: Unsigned PR vs Positive Degree\n",
    "sc1 = axes[0, 0].scatter(\n",
    "    results['degree_pos'],\n",
    "    results['pagerank_unsigned'],\n",
    "    c=results['pagerank_signed'],\n",
    "    cmap=cmap,\n",
    "    norm=norm,\n",
    "    edgecolor='black',\n",
    "    s=100\n",
    ")\n",
    "axes[0, 0].set_title(\"Unsigned PR vs Positive Degree\")\n",
    "axes[0, 0].set_xlabel(\"Positive Degree\")\n",
    "axes[0, 0].set_ylabel(\"Unsigned PageRank\")\n",
    "\n",
    "# Plot 2: Signed PR vs Positive Degree\n",
    "sc2 = axes[0, 1].scatter(\n",
    "    results['degree_pos'],\n",
    "    results['pagerank_signed'],\n",
    "    c=results['pagerank_signed'],\n",
    "    cmap=cmap,\n",
    "    norm=norm,\n",
    "    edgecolor='black',\n",
    "    s=100\n",
    ")\n",
    "axes[0, 1].set_title(\"Signed PR vs Positive Degree\")\n",
    "axes[0, 1].set_xlabel(\"Positive Degree\")\n",
    "axes[0, 1].set_ylabel(\"Signed PageRank\")\n",
    "\n",
    "# Plot 3: Unsigned PR vs Negative Degree\n",
    "sc3 = axes[1, 0].scatter(\n",
    "    results['degree_neg'],\n",
    "    results['pagerank_unsigned'],\n",
    "    c=results['pagerank_signed'],\n",
    "    cmap=cmap,\n",
    "    norm=norm,\n",
    "    edgecolor='black',\n",
    "    s=100\n",
    ")\n",
    "axes[1, 0].set_title(\"Unsigned PR vs Negative Degree\")\n",
    "axes[1, 0].set_xlabel(\"Negative Degree\")\n",
    "axes[1, 0].set_ylabel(\"Unsigned PageRank\")\n",
    "\n",
    "# Plot 4: Signed PR vs Negative Degree\n",
    "sc4 = axes[1, 1].scatter(\n",
    "    results['degree_neg'],\n",
    "    results['pagerank_signed'],\n",
    "    c=results['pagerank_signed'],\n",
    "    cmap=cmap,\n",
    "    norm=norm,\n",
    "    edgecolor='black',\n",
    "    s=100\n",
    ")\n",
    "axes[1, 1].set_title(\"Signed PR vs Negative Degree\")\n",
    "axes[1, 1].set_xlabel(\"Negative Degree\")\n",
    "axes[1, 1].set_ylabel(\"Signed PageRank\")\n",
    "\n",
    "# Colorbar outside the plots\n",
    "cbar_ax = fig.add_axes([0.93, 0.15, 0.02, 0.7])\n",
    "cbar = fig.colorbar(sc4, cax=cbar_ax)\n",
    "cbar.set_label(\"Signed PageRank\")\n",
    "\n",
    "plt.tight_layout(rect=[0, 0, 0.9, 1])  # leave room for colorbar\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "715bbfab",
   "metadata": {},
   "source": [
    "#### Pearson's and Spearman’s Rank Correlation Coefficient\n",
    "\n",
    "Calculate Pearsonr and Spearmanr to estiamte the correlation between pos/neg degree and signed/unsigned PR.\n",
    "For small sample sizes with Spearman, consider performing a permutation test instead of relying on the asymptotic p-value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "946dd69c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import pearsonr, spearmanr, permutation_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d640788c",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_labels = []\n",
    "data = []\n",
    "\n",
    "columns = pd.MultiIndex.from_tuples([\n",
    "    (\"Pearson\", \"r\"), (\"Pearson\", \"p\"),\n",
    "    (\"Spearman\", \"r\"), (\"Spearman\", \"p_asymptotic\"), (\"Spearman\", \"p_empirical\"),\n",
    "])\n",
    "\n",
    "for column1 in [\"pagerank_unsigned\", \"pagerank_signed\"]:\n",
    "    for column2 in [\"degree\", \"degree_pos\", \"degree_neg\"]:\n",
    "        if column1 != column2:\n",
    "            value_name = f\"{column1}+{column2}\"\n",
    "            x = results[column1].values\n",
    "            y = results[column2].values\n",
    "\n",
    "            # Pearson correlation\n",
    "            pearson_r, pearson_p = pearsonr(x, y)\n",
    "\n",
    "            # Spearman correlation (asymptotic)\n",
    "            spearman_result = spearmanr(x, y)\n",
    "            spearman_r = spearman_result.statistic\n",
    "            spearman_p_asymptotic = spearman_result.pvalue\n",
    "\n",
    "            # Permutation test for empirical p-value\n",
    "            def stat_fn(x_perm):\n",
    "                return spearmanr(x_perm, y).statistic\n",
    "\n",
    "            perm_result = permutation_test(\n",
    "                (x,),\n",
    "                stat_fn,\n",
    "                permutation_type='pairings',\n",
    "                alternative='two-sided',\n",
    "                n_resamples=10_000,  # increase for precision\n",
    "                random_state=42\n",
    "            )\n",
    "            spearman_p_empirical = perm_result.pvalue\n",
    "\n",
    "            data.append([\n",
    "                pearson_r, pearson_p,\n",
    "                spearman_r, spearman_p_asymptotic, spearman_p_empirical\n",
    "            ])\n",
    "            row_labels.append(value_name)\n",
    "\n",
    "df_corr = pd.DataFrame(data, index=row_labels, columns=columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c16a8bc0-a677-418c-8797-413e660994c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(2, 2, figsize=(15, 12))\n",
    "\n",
    "# Plot 1: pagerank_unsigned vs degree_pos\n",
    "column1 = \"pagerank_unsigned\"\n",
    "column2 = \"degree_pos\"\n",
    "r_value, p_value = pearsonr(results[column1], results[column2])\n",
    "\n",
    "sns.regplot(\n",
    "    x=column1,\n",
    "    y=column2,\n",
    "    data=results,\n",
    "    ci=None,\n",
    "    line_kws={\"color\": \"salmon\"},\n",
    "    scatter_kws={\"alpha\": 0.6},\n",
    "    ax=axes[0, 0]\n",
    ")\n",
    "axes[0, 0].set_title(f'Pearson r = {r_value:.4f}, p = {p_value:.4g}', fontsize=14, fontweight='bold')\n",
    "axes[0, 0].set_xlabel(column1)\n",
    "axes[0, 0].set_ylabel(column2)\n",
    "axes[0, 0].grid(True)\n",
    "\n",
    "# Plot 2: pagerank_unsigned vs degree_neg\n",
    "column1 = \"pagerank_unsigned\"\n",
    "column2 = \"degree_neg\"\n",
    "r_value, p_value = pearsonr(results[column1], results[column2])\n",
    "\n",
    "sns.regplot(\n",
    "    x=column1,\n",
    "    y=column2,\n",
    "    data=results,\n",
    "    ci=None,\n",
    "    line_kws={\"color\": \"salmon\"},\n",
    "    scatter_kws={\"alpha\": 0.6},\n",
    "    ax=axes[0, 1]\n",
    ")\n",
    "axes[0, 1].set_title(f'Pearson r = {r_value:.4f}, p = {p_value:.4g}', fontsize=14, fontweight='bold')\n",
    "axes[0, 1].set_xlabel(column1)\n",
    "axes[0, 1].set_ylabel(column2)\n",
    "axes[0, 1].grid(True)\n",
    "\n",
    "# Plot 3: pagerank_signed vs degree_pos\n",
    "column1 = \"pagerank_signed\"\n",
    "column2 = \"degree_pos\"\n",
    "r_value, p_value = pearsonr(results[column1], results[column2])\n",
    "\n",
    "sns.regplot(\n",
    "    x=column1,\n",
    "    y=column2,\n",
    "    data=results,\n",
    "    ci=None,\n",
    "    line_kws={\"color\": \"salmon\"},\n",
    "    scatter_kws={\"alpha\": 0.6},\n",
    "    ax=axes[1, 0]\n",
    ")\n",
    "axes[1, 0].set_title(f'Pearson r = {r_value:.4f}, p = {p_value:.4g}', fontsize=14, fontweight='bold')\n",
    "axes[1, 0].set_xlabel(column1)\n",
    "axes[1, 0].set_ylabel(column2)\n",
    "axes[1, 0].grid(True)\n",
    "\n",
    "# Plot 4: pagerank_signed vs degree_neg\n",
    "column1 = \"pagerank_signed\"\n",
    "column2 = \"degree_neg\"\n",
    "r_value, p_value = pearsonr(results[column1], results[column2])\n",
    "\n",
    "sns.regplot(\n",
    "    x=column1,\n",
    "    y=column2,\n",
    "    data=results,\n",
    "    ci=None,\n",
    "    line_kws={\"color\": \"salmon\"},\n",
    "    scatter_kws={\"alpha\": 0.6},\n",
    "    ax=axes[1, 1]\n",
    ")\n",
    "axes[1, 1].set_title(f'Pearson r = {r_value:.4f}, p = {p_value:.4g}', fontsize=14, fontweight='bold')\n",
    "axes[1, 1].set_xlabel(column1)\n",
    "axes[1, 1].set_ylabel(column2)\n",
    "axes[1, 1].grid(True)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d612a787-5c40-47c5-bdd4-26bceb9d4cf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(2, 2, figsize=(15, 12))\n",
    "\n",
    "# Plot 1: pagerank_unsigned vs degree_pos\n",
    "column1 = \"pagerank_unsigned\"\n",
    "column2 = \"degree_pos\"\n",
    "r_value, p_value = spearmanr(results[column1], results[column2])\n",
    "ranked_data = results[[column1, column2]].rank()\n",
    "\n",
    "sns.regplot(\n",
    "    x=column1,\n",
    "    y=column2,\n",
    "    data=ranked_data,\n",
    "    ci=None,\n",
    "    line_kws={\"color\": \"orange\"},\n",
    "    scatter_kws={\"alpha\": 0.6},\n",
    "    ax=axes[0, 0]\n",
    ")\n",
    "axes[0, 0].set_title(f'Spearman r = {r_value:.4f}, p = {p_value:.4g}')\n",
    "axes[0, 0].set_xlabel(f'Rank of {column1}')\n",
    "axes[0, 0].set_ylabel(f'Rank of {column2}')\n",
    "axes[0, 0].grid(True)\n",
    "\n",
    "# Plot 2: pagerank_unsigned vs degree_neg\n",
    "column1 = \"pagerank_unsigned\"\n",
    "column2 = \"degree_neg\"\n",
    "r_value, p_value = spearmanr(results[column1], results[column2])\n",
    "ranked_data = results[[column1, column2]].rank()\n",
    "\n",
    "sns.regplot(\n",
    "    x=column1,\n",
    "    y=column2,\n",
    "    data=ranked_data,\n",
    "    ci=None,\n",
    "    line_kws={\"color\": \"orange\"},\n",
    "    scatter_kws={\"alpha\": 0.6},\n",
    "    ax=axes[0, 1]\n",
    ")\n",
    "axes[0, 1].set_title(f'Spearman r = {r_value:.4f}, p = {p_value:.4g}')\n",
    "axes[0, 1].set_xlabel(f'Rank of {column1}')\n",
    "axes[0, 1].set_ylabel(f'Rank of {column2}')\n",
    "axes[0, 1].grid(True)\n",
    "\n",
    "# Plot 3: pagerank_signed vs degree_pos\n",
    "column1 = \"pagerank_signed\"\n",
    "column2 = \"degree_pos\"\n",
    "r_value, p_value = spearmanr(results[column1], results[column2])\n",
    "ranked_data = results[[column1, column2]].rank()\n",
    "\n",
    "sns.regplot(\n",
    "    x=column1,\n",
    "    y=column2,\n",
    "    data=ranked_data,\n",
    "    ci=None,\n",
    "    line_kws={\"color\": \"orange\"},\n",
    "    scatter_kws={\"alpha\": 0.6},\n",
    "    ax=axes[1, 0]\n",
    ")\n",
    "axes[1, 0].set_title(f'Spearman r = {r_value:.4f}, p = {p_value:.4g}')\n",
    "axes[1, 0].set_xlabel(f'Rank of {column1}')\n",
    "axes[1, 0].set_ylabel(f'Rank of {column2}')\n",
    "axes[1, 0].grid(True)\n",
    "\n",
    "# Plot 4: pagerank_signed vs degree_neg\n",
    "column1 = \"pagerank_signed\"\n",
    "column2 = \"degree_neg\"\n",
    "r_value, p_value = spearmanr(results[column1], results[column2])\n",
    "ranked_data = results[[column1, column2]].rank()\n",
    "\n",
    "sns.regplot(\n",
    "    x=column1,\n",
    "    y=column2,\n",
    "    data=ranked_data,\n",
    "    ci=None,\n",
    "    line_kws={\"color\": \"orange\"},\n",
    "    scatter_kws={\"alpha\": 0.6},\n",
    "    ax=axes[1, 1]\n",
    ")\n",
    "axes[1, 1].set_title(f'Spearman r = {r_value:.4f}, p = {p_value:.4g}')\n",
    "axes[1, 1].set_xlabel(f'Rank of {column1}')\n",
    "axes[1, 1].set_ylabel(f'Rank of {column2}')\n",
    "axes[1, 1].grid(True)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3c4697d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flatten multi-index columns\n",
    "df_flat = df_corr.copy()\n",
    "df_flat.columns = ['_'.join(col).strip() for col in df_flat.columns.values]\n",
    "df_flat = df_flat.reset_index().rename(columns={\"index\": \"Variable Pair\"})\n",
    "\n",
    "df_long = df_flat.melt(\n",
    "    id_vars=\"Variable Pair\",\n",
    "    value_vars=[\"Pearson_r\", \"Spearman_r\"],\n",
    "    var_name=\"Correlation Type\",\n",
    "    value_name=\"Correlation Coefficient\"\n",
    ")\n",
    "\n",
    "plt.figure(figsize=(12, 6))\n",
    "sns.barplot(\n",
    "    data=df_long,\n",
    "    x=\"Variable Pair\",\n",
    "    y=\"Correlation Coefficient\",\n",
    "    hue=\"Correlation Type\",\n",
    "    palette=[\"#1f77b4\", \"#ff7f0e\"]\n",
    ")\n",
    "\n",
    "for i, row in df_flat.iterrows():\n",
    "    spearman_p = row[\"Spearman_p_empirical\"]\n",
    "    pearson_p = row[\"Pearson_p\"]\n",
    "    plt.text(i, 1.05, f'Spearman_p={spearman_p:.3f}', ha='center', va='bottom', fontsize=9, color='black')\n",
    "    plt.text(i, 1.12, f'Pearson_p={pearson_p:.3f}', ha='center', va='bottom', fontsize=9, color='black')\n",
    "\n",
    "\n",
    "plt.axhline(0, color='gray', linestyle='--')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.title(\"Comparison of Pearson and Spearman Correlation Coefficients\", pad=30)\n",
    "plt.tight_layout()\n",
    "plt.grid(axis='y', linestyle='--', alpha=0.5)\n",
    "plt.legend(title=\"Correlation Type\")\n",
    "plt.savefig(os.path.join(\"figures\",'pearson_spearman_correlation.jpg'))\n",
    "plt.show()\n",
    "\n",
    "\n",
    "df_pvals_long = df_flat.melt(\n",
    "    id_vars=\"Variable Pair\",\n",
    "    value_vars=[\"Spearman_p_asymptotic\", \"Spearman_p_empirical\"],\n",
    "    var_name=\"P-value Type\",\n",
    "    value_name=\"P-value\"\n",
    ")\n",
    "\n",
    "plt.figure(figsize=(12, 6))\n",
    "sns.barplot(\n",
    "    data=df_pvals_long,\n",
    "    x=\"Variable Pair\",\n",
    "    y=\"P-value\",\n",
    "    hue=\"P-value Type\",\n",
    "    palette=[\"#2ca02c\", \"#d62728\"]\n",
    ")\n",
    "\n",
    "plt.axhline(0, color='gray', linestyle='--')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.title(\"Comparison of Pearson and Spearman Correlation Coefficients\", pad=20)\n",
    "plt.legend(title=\"P-value Type\")\n",
    "plt.tight_layout()\n",
    "plt.grid(axis='y', linestyle='--', alpha=0.5)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb4970f3-1594-4a29-a1a9-bc771c046475",
   "metadata": {},
   "source": [
    "#### Histogram\n",
    "Comparison of signed PageRank and Prestige and positive/negative degree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13497073-8577-4ee9-b91a-55b32ebea0d2",
   "metadata": {},
   "source": [
    "##### PageRank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c18a9960-778c-4bab-835b-4b08a7e89590",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(2, 2, figsize=(20, 15))\n",
    "\n",
    "# mask for positive / negative signed PR\n",
    "pos_PR_mask = results['pagerank_signed'] > 0\n",
    "neg_PR_mask = results['pagerank_signed'] < 0\n",
    "\n",
    "# mask for positive / negative degree\n",
    "pos_degree_mask = results['degree_pos'] > 0\n",
    "neg_degree_mask = results['degree_neg'] > 0\n",
    "\n",
    "# mask to filter nodes based on which degree (positive or negative) is higher, only consider the higher degree\n",
    "pos_degree_mask_filtered = results['degree_pos'] > results['degree_neg']\n",
    "neg_degree_mask_filtered = results['degree_neg'] > results['degree_pos']\n",
    "\n",
    "\n",
    "\n",
    "# Plot 1: Degree by Signed PR (Unfiltered)\n",
    "pos_PR_combined_degrees_unfiltered = pd.concat([results.loc[pos_PR_mask, 'degree_pos'], -results.loc[pos_PR_mask, 'degree_neg']])\n",
    "neg_PR_combined_degrees_unfiltered = pd.concat([results.loc[neg_PR_mask, 'degree_pos'], -results.loc[neg_PR_mask, 'degree_neg']])\n",
    "\n",
    "all_degrees_unfiltered = pd.concat([pos_PR_combined_degrees_unfiltered, neg_PR_combined_degrees_unfiltered])\n",
    "min_val_unfiltered = all_degrees_unfiltered.min()\n",
    "max_val_unfiltered = all_degrees_unfiltered.max()\n",
    "bins_unfiltered = np.linspace(min_val_unfiltered, max_val_unfiltered, 30)\n",
    "\n",
    "axes[0, 0].hist(\n",
    "    [pos_PR_combined_degrees_unfiltered, neg_PR_combined_degrees_unfiltered],\n",
    "    bins=bins_unfiltered,\n",
    "    color=['skyblue', 'salmon'],\n",
    "    alpha=0.8,\n",
    "    label=['Positive Signed PR', 'Negative Signed PR'],\n",
    "    edgecolor='black'\n",
    ")\n",
    "axes[0, 0].axvline(0, color='black', linestyle='--', linewidth=1.5)\n",
    "axes[0, 0].set_title(\"Distribution of Degree by Signed PR\", fontsize=14, fontweight='bold')\n",
    "axes[0, 0].set_xlabel(\"Degree (Positive and Negative)\")\n",
    "axes[0, 0].set_ylabel(\"Frequency (Number of Nodes)\")\n",
    "axes[0, 0].legend(frameon=True, fontsize=10)\n",
    "axes[0, 0].grid(axis='y', linestyle='--', alpha=0.6)\n",
    "\n",
    "# Plot 2: Degree by Signed PR (Filtered by Dominant Degree)\n",
    "pos_pr_combined_filtered = pd.concat([results.loc[pos_PR_mask & pos_degree_mask_filtered, 'degree_pos'], -results.loc[pos_PR_mask & neg_degree_mask_filtered, 'degree_neg']])\n",
    "neg_pr_combined_filtered = pd.concat([results.loc[neg_PR_mask & pos_degree_mask_filtered, 'degree_pos'], -results.loc[neg_PR_mask & neg_degree_mask_filtered, 'degree_neg']])\n",
    "\n",
    "all_degrees_pr_filtered = pd.concat([pos_pr_combined_filtered, neg_pr_combined_filtered])\n",
    "min_val_pr_filtered = all_degrees_pr_filtered.min()\n",
    "max_val_pr_filtered = all_degrees_pr_filtered.max()\n",
    "bins_pr_filtered = np.linspace(min_val_pr_filtered, max_val_pr_filtered, 30)\n",
    "\n",
    "axes[0, 1].hist(\n",
    "    [pos_pr_combined_filtered, neg_pr_combined_filtered],\n",
    "    bins=bins_pr_filtered,\n",
    "    color=['skyblue', 'salmon'],\n",
    "    alpha=0.8,\n",
    "    label=['Positive Signed PR', 'Negative Signed PR'],\n",
    "    edgecolor='black'\n",
    ")\n",
    "axes[0, 1].axvline(0, color='black', linestyle='--', linewidth=1.5)\n",
    "axes[0, 1].set_title(\"Distribution of Degree (Filtered by Dominant Degree) by Signed PR\", fontsize=14, fontweight='bold')\n",
    "axes[0, 1].set_xlabel(\"Degree (Positive and Negative)\")\n",
    "axes[0, 1].set_ylabel(\"Frequency (Number of Nodes)\")\n",
    "axes[0, 1].legend(frameon=True, fontsize=10)\n",
    "axes[0, 1].grid(axis='y', linestyle='--', alpha=0.6)\n",
    "\n",
    "# Plot 3: Signed PR by Degree Type (Unfiltered)\n",
    "axes[1, 0].hist(\n",
    "    [results.loc[pos_degree_mask, 'pagerank_signed'], results.loc[neg_degree_mask, 'pagerank_signed']],\n",
    "    bins=20,\n",
    "    color=['skyblue', 'salmon'],\n",
    "    alpha=0.8,\n",
    "    label=['Positive Degree', 'Negative Degree'],\n",
    "    edgecolor='black'\n",
    ")\n",
    "axes[1, 0].axvline(0, color='black', linestyle='--', linewidth=1.5)\n",
    "axes[1, 0].set_title(\"Distribution of Signed PR by Degree Type (Unfiltered)\", fontsize=14, fontweight='bold')\n",
    "axes[1, 0].set_xlabel(\"Signed PageRank\")\n",
    "axes[1, 0].set_ylabel(\"Frequency (Number of Nodes)\")\n",
    "axes[1, 0].legend(frameon=True, fontsize=10)\n",
    "axes[1, 0].grid(axis='y', linestyle='--', alpha=0.6)\n",
    "\n",
    "# Plot 4: Signed PR by Degree Type (Filtered by Dominant Degree)\n",
    "axes[1, 1].hist(\n",
    "    [results.loc[pos_degree_mask_filtered, 'pagerank_signed'], results.loc[neg_degree_mask_filtered, 'pagerank_signed']],\n",
    "    bins=20,\n",
    "    color=['skyblue', 'salmon'],\n",
    "    alpha=0.8,\n",
    "    label=['Positive Degree', 'Negative Degree'],\n",
    "    edgecolor='black'\n",
    ")\n",
    "axes[1, 1].axvline(0, color='black', linestyle='--', linewidth=1.5)\n",
    "axes[1, 1].set_title(\"Distribution of Signed PR by Degree Type (Filtered by Dominant Degree)\", fontsize=14, fontweight='bold')\n",
    "axes[1, 1].set_xlabel(\"Signed PageRank\")\n",
    "axes[1, 1].set_ylabel(\"Frequency (Number of Nodes)\")\n",
    "axes[1, 1].legend(frameon=True, fontsize=10)\n",
    "axes[1, 1].grid(axis='y', linestyle='--', alpha=0.6)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(os.path.join(\"figures\", 'histogram_combined_degree_PR.png'))\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22d7bfa2-81be-4d36-b732-f9f64b18d6aa",
   "metadata": {},
   "source": [
    "##### Prestige"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d30721f6-40d5-43f0-84c6-2d6efdf9fd0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(2, 2, figsize=(20, 15))\n",
    "\n",
    "# mask for positive / negative prestige\n",
    "pos_prestige_mask = results['prestige'] > 0\n",
    "neg_prestige_mask = results['prestige'] < 0\n",
    "\n",
    "# mask for positive / negative degree\n",
    "pos_degree_mask = results['degree_pos'] > 0\n",
    "neg_degree_mask = results['degree_neg'] > 0\n",
    "\n",
    "# mask to filter nodes based on which degree (positive or negative) is higher, only consider the higher degree\n",
    "pos_degree_mask_filtered = results['degree_pos'] > results['degree_neg']\n",
    "neg_degree_mask_filtered = results['degree_neg'] > results['degree_pos']\n",
    "\n",
    "\n",
    "# Plot 1: Combined Degree by Prestige (Unfiltered)\n",
    "pos_prestige_combined_degrees_unfiltered = pd.concat([results.loc[pos_prestige_mask, 'degree_pos'], -results.loc[pos_prestige_mask, 'degree_neg']])\n",
    "neg_prestige_combined_degrees_unfiltered = pd.concat([results.loc[neg_prestige_mask, 'degree_pos'], -results.loc[neg_prestige_mask, 'degree_neg']])\n",
    "\n",
    "all_degrees_unfiltered = pd.concat([pos_prestige_combined_degrees_unfiltered, neg_prestige_combined_degrees_unfiltered])\n",
    "min_val_unfiltered = all_degrees_unfiltered.min()\n",
    "max_val_unfiltered = all_degrees_unfiltered.max()\n",
    "bins_unfiltered = np.linspace(min_val_unfiltered, max_val_unfiltered, 30)\n",
    "\n",
    "axes[0, 0].hist(\n",
    "    [pos_prestige_combined_degrees_unfiltered, neg_prestige_combined_degrees_unfiltered],\n",
    "    bins=bins_unfiltered,\n",
    "    color=['skyblue', 'salmon'],\n",
    "    alpha=0.8,\n",
    "    label=['Positive Prestige', 'Negative Prestige'],\n",
    "    edgecolor='black'\n",
    ")\n",
    "axes[0, 0].axvline(0, color='black', linestyle='--', linewidth=1.5)\n",
    "axes[0, 0].set_title(\"Distribution of Degree by Prestige (Unfiltered)\", fontsize=14, fontweight='bold')\n",
    "axes[0, 0].set_xlabel(\"Degree (Positive and Negative)\")\n",
    "axes[0, 0].set_ylabel(\"Frequency (Number of Nodes)\")\n",
    "axes[0, 0].legend(frameon=True, fontsize=10)\n",
    "axes[0, 0].grid(axis='y', linestyle='--', alpha=0.6)\n",
    "\n",
    "# Plot 2: Combined Degree by Prestige (Filtered)\n",
    "pos_prestige_combined_filtered = pd.concat([results.loc[pos_prestige_mask & pos_degree_mask_filtered, 'degree_pos'], -results.loc[pos_prestige_mask & neg_degree_mask_filtered, 'degree_neg']])\n",
    "neg_prestige_combined_filtered = pd.concat([results.loc[neg_prestige_mask & pos_degree_mask_filtered, 'degree_pos'], -results.loc[neg_prestige_mask & neg_degree_mask_filtered, 'degree_neg']])\n",
    "\n",
    "all_degrees_filtered = pd.concat([pos_prestige_combined_filtered, neg_prestige_combined_filtered])\n",
    "min_val_filtered = all_degrees_filtered.min()\n",
    "max_val_filtered = all_degrees_filtered.max()\n",
    "bins_filtered = np.linspace(min_val_filtered, max_val_filtered, 30)\n",
    "\n",
    "axes[0, 1].hist(\n",
    "    [pos_prestige_combined_filtered, neg_prestige_combined_filtered],\n",
    "    bins=bins_filtered,\n",
    "    color=['skyblue', 'salmon'],\n",
    "    alpha=0.8,\n",
    "    label=['Positive Prestige', 'Negative Prestige'],\n",
    "    edgecolor='black'\n",
    ")\n",
    "axes[0, 1].axvline(0, color='black', linestyle='--', linewidth=1.5)\n",
    "axes[0, 1].set_title(\"Distribution of Degree by Prestige (Filtered by Dominant Degree)\", fontsize=14, fontweight='bold')\n",
    "axes[0, 1].set_xlabel(\"Degree (Positive and Negative)\")\n",
    "axes[0, 1].set_ylabel(\"Frequency (Number of Nodes)\")\n",
    "axes[0, 1].legend(frameon=True, fontsize=10)\n",
    "axes[0, 1].grid(axis='y', linestyle='--', alpha=0.6)\n",
    "\n",
    "# Plot 3: Prestige by Degree Type (Unfiltered)\n",
    "axes[1, 0].hist(\n",
    "    [results.loc[pos_degree_mask, 'prestige'], results.loc[neg_degree_mask, 'prestige']],\n",
    "    bins=20,\n",
    "    color=['skyblue', 'salmon'],\n",
    "    alpha=0.8,\n",
    "    label=['Positive Degree', 'Negative Degree'],\n",
    "    edgecolor='black'\n",
    ")\n",
    "axes[1, 0].axvline(0, color='black', linestyle='--', linewidth=1.5)\n",
    "axes[1, 0].set_title(\"Distribution of Prestige by Degree Type (Unfiltered)\", fontsize=14, fontweight='bold')\n",
    "axes[1, 0].set_xlabel(\"Prestige\")\n",
    "axes[1, 0].set_ylabel(\"Frequency (Number of Nodes)\")\n",
    "axes[1, 0].legend(frameon=True, fontsize=10)\n",
    "axes[1, 0].grid(axis='y', linestyle='--', alpha=0.6)\n",
    "\n",
    "# Plot 4: Prestige by Degree Type (Filtered)\n",
    "axes[1, 1].hist(\n",
    "    [results.loc[pos_degree_mask_filtered, 'prestige'], results.loc[neg_degree_mask_filtered, 'prestige']],\n",
    "    bins=20,\n",
    "    color=['skyblue', 'salmon'],\n",
    "    alpha=0.8,\n",
    "    label=['Positive Degree', 'Negative Degree'],\n",
    "    edgecolor='black'\n",
    ")\n",
    "axes[1, 1].axvline(0, color='black', linestyle='--', linewidth=1.5)\n",
    "axes[1, 1].set_title(\"Distribution of Prestige by Degree Type (Filtered by Dominant Degree)\", fontsize=14, fontweight='bold')\n",
    "axes[1, 1].set_xlabel(\"Prestige\")\n",
    "axes[1, 1].set_ylabel(\"Frequency (Number of Nodes)\")\n",
    "axes[1, 1].legend(frameon=True, fontsize=10)\n",
    "axes[1, 1].grid(axis='y', linestyle='--', alpha=0.6)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(os.path.join(\"figures\", 'histogram_combined_degree_prestige.png'))\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6108a52-cd19-44f8-a81d-066913261bed",
   "metadata": {},
   "source": [
    "### Degree vs Centrality Metric for each node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43baf6d4-5e82-4048-ba45-622e9ca49e07",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_signed_centrality(df, metric_col, title):\n",
    "    \"\"\"\n",
    "    Generates bar plots for nodes, split by whether their centrality metric is positive or negative.\n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame): The DataFrame containing the node data.\n",
    "        metric_col (str): The name of the column containing the signed centrality metric.\n",
    "        title (str): The title for the plot.\n",
    "    \"\"\"\n",
    "    # Filter out rows where the metric value is NaN\n",
    "    filtered_df = df.dropna(subset=[metric_col])\n",
    "\n",
    "    # Separate nodes with positive and negative scores for the given metric\n",
    "    positive_nodes = filtered_df[filtered_df[metric_col] >= 0]\n",
    "    negative_nodes = filtered_df[filtered_df[metric_col] < 0]\n",
    "    \n",
    "    # Define colors for positive and negative degrees\n",
    "    pos_degree_color = '#4CAF50'  # A nice green for positive degree\n",
    "    neg_degree_color = '#F44336'  # A vibrant red for negative degree\n",
    "    bar_width = 0.35\n",
    "\n",
    "    # Plot for nodes with a positive metric score\n",
    "    fig_pos, ax_pos = plt.subplots(figsize=(15, 8))\n",
    "    ind_pos = np.arange(len(positive_nodes.index))\n",
    "    ax_pos.bar(ind_pos - bar_width/2, positive_nodes['degree_pos'], bar_width, label='Positive Degree',\n",
    "               color=pos_degree_color)\n",
    "    ax_pos.bar(ind_pos + bar_width/2, positive_nodes['degree_neg'], bar_width, label='Negative Degree',\n",
    "               color=neg_degree_color)\n",
    "    ax_pos.set_ylabel('Degree Value', fontsize=12)\n",
    "    ax_pos.set_xlabel('Node', fontsize=12)\n",
    "    ax_pos.set_title(f'{title} (Positive Scores)', fontsize=16)\n",
    "    ax_pos.set_xticks(ind_pos)\n",
    "    ax_pos.set_xticklabels(positive_nodes.index, rotation=45, ha='right', fontsize=10)\n",
    "    ax_pos.legend()\n",
    "    ax_pos.yaxis.grid(True, linestyle='--', alpha=0.6)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(os.path.join(\"figures\", f'{title}_positive_scores.png'))\n",
    "    plt.show()\n",
    "\n",
    "    # Plot for nodes with a negative metric score\n",
    "    fig_neg, ax_neg = plt.subplots(figsize=(15, 8))\n",
    "    ind_neg = np.arange(len(negative_nodes.index))\n",
    "    ax_neg.bar(ind_neg - bar_width/2, negative_nodes['degree_pos'], bar_width, label='Positive Degree',\n",
    "               color=pos_degree_color)\n",
    "    ax_neg.bar(ind_neg + bar_width/2, negative_nodes['degree_neg'], bar_width, label='Negative Degree',\n",
    "               color=neg_degree_color)\n",
    "    ax_neg.set_ylabel('Degree Value', fontsize=12)\n",
    "    ax_neg.set_xlabel('Node', fontsize=12)\n",
    "    ax_neg.set_title(f'{title} (Negative Scores)', fontsize=16)\n",
    "    ax_neg.set_xticks(ind_neg)\n",
    "    ax_neg.set_xticklabels(negative_nodes.index, rotation=45, ha='right', fontsize=10)\n",
    "    ax_neg.legend()\n",
    "    ax_neg.yaxis.grid(True, linestyle='--', alpha=0.6)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(os.path.join(\"figures\", f'{title}_negative_scores.png'))\n",
    "    plt.show()\n",
    "\n",
    "# Generate the plots for signed PageRank and prestige\n",
    "plot_signed_centrality(results, 'pagerank_signed', 'Positive vs. Negative Degree for Signed PageRank')\n",
    "plot_signed_centrality(results, 'prestige', 'Positive vs. Negative Degree for Signed Prestige')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecde3ee1-120b-4913-8fc8-69d523f51c9b",
   "metadata": {},
   "source": [
    "### Niche plots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f38abed6-8006-4ec2-875b-f3be8f68c0b3",
   "metadata": {},
   "source": [
    "And also visualize niches in slices (spots are colored by the niche to which the cell type with highest proportion belongs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "732788cf-dd8c-4663-bb48-cc603df5a59e",
   "metadata": {},
   "source": [
    "Now we will visualize the **niches** in the **slices** coloring the Visium spots according to the niche of the cell type with the highest proportion. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c65a6826-960d-4305-b0db-6834dbf21392",
   "metadata": {},
   "source": [
    "These are a couple **myogenic** slices, which will be at the **top** panels of the next figure:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5956ee75-dab7-4427-b96c-41d23bc8fb15",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(2, 3, figsize=(21, 7))\n",
    "plt.close(fig)\n",
    "for idu,smpl in enumerate(list(sampleTypesDF['sample'][sampleTypesDF['sampleType']=='myogenic'][0:3])):  \n",
    "    _ = nichesphere.coloc.spatialNichePlot(adata=mudata['visium'][mudata['visium'].obs.patient_region_id==smpl].copy(), \n",
    "                                       CTprobs=CTprops.loc[spotSamples.index[spotSamples==smpl]],    #dataframe of cell type probabilities per spot\n",
    "                                       cell_types=mudata['sc'].obs.cell_subtype2,                    #categorical series of cell types\n",
    "                                       nicheDF=niches_df, \n",
    "                                       spot_size=0.015, \n",
    "                                       niche_colors=niche_cols,                                      #series of colors with niche names as indexes\n",
    "                                       legend_fontsize=7, save_name='_'+smpl+'.pdf',ax=axes[0][idu])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8a60cf1-1dc9-443f-994f-ef3f1f446727",
   "metadata": {},
   "source": [
    "And a couple **ischemic** slices, which will be at the **bottom** panels of the next figure:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b94959a-e5d2-47d9-aa83-e23880d806c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "for idu,smpl in enumerate(list(sampleTypesDF['sample'][sampleTypesDF['sampleType']=='ischemic'][0:3])):  \n",
    "    _ = nichesphere.coloc.spatialNichePlot(adata=mudata['visium'][mudata['visium'].obs.patient_region_id==smpl].copy(), \n",
    "                                       CTprobs=CTprops.loc[spotSamples.index[spotSamples==smpl]], \n",
    "                                       cell_types=mudata['sc'].obs.cell_subtype2, \n",
    "                                       nicheDF=niches_df, \n",
    "                                       spot_size=0.015, \n",
    "                                       niche_colors=niche_cols, \n",
    "                                       legend_fontsize=7, \n",
    "                                       save_name='_'+smpl+'.pdf',ax=axes[1][idu])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "762e9663-1ca4-41f5-a83e-2e4b7af95e9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig.tight_layout()\n",
    "fig"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1ef977c-ac57-4131-8e2a-92a5e42bd69e",
   "metadata": {},
   "source": [
    "For further analysis, like differential communication: https://nichesphere.readthedocs.io/en/latest/notebooks/Nichesphere_tutorial_MIvisium_comm.html \n",
    "\n",
    ", we will need the correspondence data between cell pairs and niche pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1a26959-2014-4349-bd92-ba0803ff43d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairCatDFdir=nichesphere.tl.get_pairCatDFdir(niches_df)\n",
    "pairCatDFdir.to_csv('pairCatDFdir_MIvisium_louvain.csv')\n",
    "pairCatDFdir.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8914d2fd-d638-4259-bba3-81e1bb7bc740",
   "metadata": {},
   "source": [
    "We will also need a filtering object **(colocFilt)** indicating which cell pairs are differentially co-localized to filter the communication data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77a453e7-6ec9-4d88-a2d3-e618c98a25f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get data of cells present in the adjacency matrix\n",
    "pairCatDF_filter=[(pairCatDFdir.cell_pairs.str.split('->')[i][0] in HMsimm.index)&\n",
    "                  (pairCatDFdir.cell_pairs.str.split('->')[i][1] in HMsimm.index) for i in pairCatDFdir.index]\n",
    "pairCatDFdir_filt=pairCatDFdir[pairCatDF_filter]\n",
    "oneCTints_filt=oneCTints[[i.split('-')[0] in HMsimm.index for i in oneCTints]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "535ecf21-84b7-4c0e-a81e-658bd279238f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get data to flag differentially co-localized cell pairs in the adjacency matrix\n",
    "colocFilt=nichesphere.tl.getColocFilter(pairCatDF=pairCatDFdir_filt, \n",
    "                                        adj=HMsimm, \n",
    "                                        oneCTints=oneCTints_filt.str.replace('-', '->'))\n",
    "colocFilt.to_csv('colocFilt_MIvisium_louvain.csv')\n",
    "colocFilt.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70616850-81fa-4f6b-b677-9362c9f03233",
   "metadata": {},
   "source": [
    "We will need the niche - cell type - color correspondence data, the co-localization network and nodes positions for further analysis as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "503d6b00-76ed-40f9-8c30-9ffe4b5019a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "niches_df.to_csv('niches_df_MIvisium_louvain.csv')\n",
    "nx.write_graphml_lxml(gCol, \"colocNW_MIvisium_louvain.graphml\")\n",
    "np.save('colocNW_pos.npy', pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2627a630-ba09-4568-aed4-0f86f157e3b8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nichesphere",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
